### PR TITLE
network: cleanups about Hostname= option for DHCP

### PR DESCRIPTION
**Pull request by yuwata**
Follow-up for #9766.
C.f. #9759.

### CODE REVIEW

File: src/network/networkd-network.c [hunk 1]
```diff
@@ -979,12 +980,18 @@ int config_parse_hostname(
...
+                return 0;
+        }
+
+        if (dns_name_is_valid(hn) <= 0) {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Hostname is not a valid DNS domain name, ignoring assignment: %s", rvalue);
                 return 0;
         }
 
         free(*hostname);
         *hostname = hostname_cleanup(hn);
```
Comment (keszybz) [inline src/network/networkd-network.c]: Hmm, calling `hostname_cleanup` after validation seems wrong. Shouldn't it be the other way around?

File: src/network/networkd-network.c [hunk 2]
```diff
@@ -1020,12 +1021,10 @@ int config_parse_timezone(
 
         if (!timezone_is_valid(tz, LOG_ERR)) {
                 log_syntax(unit, LOG_ERR, filename, line, 0, "Timezone is not valid, ignoring assignment: %s", rvalue);
-                free(tz);
                 return 0;
         }
 
-        free(*datap);
-        *datap = tz;
+        free_and_replace(*datap, tz);
```
Comment (keszybz) [inline src/network/networkd-network.c]: `return free_and_replace(...)` ?

File: src/network/networkd-dhcp4.c [hunk 3]
```diff
@@ -599,7 +599,14 @@ static int dhcp4_set_hostname(Link *link) {
                 hn = hostname;
         }
 
-        return sd_dhcp_client_set_hostname(link->dhcp_client, hn);
+        r = sd_dhcp_client_set_hostname(link->dhcp_client, hn);
+        if (r == -EINVAL && hostname)
+                /* Ignore error when the machine's hostname is not suitable to send in DHCP packet. */
+                log_link_info_errno(link, r, "DHCP4 CLIENT: Failed to set hostname, ignoring: %m");
```
Comment (keszybz) [inline src/network/networkd-dhcp4.c]: maybe add "... from uname, ignoring"?

File: src/network/networkd-dhcp6.c [hunk 4]
```diff
@@ -453,7 +453,14 @@ static int dhcp6_set_hostname(sd_dhcp6_client *client, Link *link) {
                 hn = hostname;
         }
 
-        return sd_dhcp6_client_set_fqdn(client, hn);
+        r = sd_dhcp6_client_set_fqdn(client, hn);
+        if (r == -EINVAL && hostname)
+                /* Ignore error when the machine's hostname is not suitable to send in DHCP packet. */
+                log_link_info_errno(link, r, "DHCP6 CLIENT: Failed to set hostname, ignoring: %m");
```
Comment (keszybz) [inline src/network/networkd-dhcp6.c]: Same here.

File: src/network/networkd-dhcp4.c (hunk 3 above)
Comment (keszybz) [reply src/network/networkd-dhcp4.c]: Or maybe " ... from kernel hostname, ignoring"?

**Comment by yuwata** [comment]
@keszybz Thank you for your review.
- Drop `hostname_cleanup()`. I think that dropping unsuitable characters from user specified `Hostname=` may causes obtaining unexpected address from DHCP server.
- Add error logs for the case that `dns_name_is_valid()` returns negative errno.

I think all other comments are addressed. PTAL.

File: src/network/networkd-dhcp4.c [hunk 5]
```diff
@@ -599,7 +599,14 @@ static int dhcp4_set_hostname(Link *link) {
                 hn = hostname;
         }
 
-        return sd_dhcp_client_set_hostname(link->dhcp_client, hn);
+        r = sd_dhcp_client_set_hostname(link->dhcp_client, hn);
+        if (r == -EINVAL && hostname)
+                /* Ignore error when the machine's hostname is not suitable to send in DHCP packet. */
+                log_link_info_errno(link, r, "DHCP4 CLIENT: Failed to set hostname from kernel hostname, ignoring: %m");
```
Comment (poettering) [inline src/network/networkd-dhcp4.c]: should be warning level, not info level i figure

File: src/network/networkd-dhcp6.c [hunk 6]
```diff
@@ -453,7 +453,14 @@ static int dhcp6_set_hostname(sd_dhcp6_client *client, Link *link) {
                 hn = hostname;
         }
 
-        return sd_dhcp6_client_set_fqdn(client, hn);
+        r = sd_dhcp6_client_set_fqdn(client, hn);
+        if (r == -EINVAL && hostname)
+                /* Ignore error when the machine's hostname is not suitable to send in DHCP packet. */
+                log_link_info_errno(link, r, "DHCP6 CLIENT: Failed to set hostname from kernel hostname, ignoring: %m");
```
Comment (poettering) [inline src/network/networkd-dhcp6.c]: info → warning

**Comment by poettering** [comment]
looks excellent, just two nitpicks about the used log level...

**Comment by yuwata** [comment]
@poettering Thank you for the review. I've adjusted the addressed log levels.

File: man/systemd.network.xml [hunk 7]
```diff
@@ -1194,25 +1194,28 @@
         <varlistentry>
           <term><varname>SendHostname=</varname></term>
           <listitem>
-            <para>When true (the default), the machine's hostname will
-            be sent to the DHCP server.</para>
+            <para>When true (the default), the machine's hostname will be sent to the DHCP server.
+            Note that the machine's hostname must consist only of 7-bit ASCII lower-case characters and
+            no spaces or dots, and be formatted as a valid DNS domain name. Otherwise, the hostname is not
+            send even if this is set to true.</para>
```
Comment (keszybz) [inline man/systemd.network.xml:1200]: is not sen**t**

Comment (keszybz) [reply man/systemd.network.xml:1200]: I fixed the typo in the merge commit.

Comment (yuwata) [reply man/systemd.network.xml:1200]: Ouch. Thanks.

**Pull request merged**